libcsf/man3/RuseAs.3
.lf 1 RuseAs.3
.\" WARNING! THIS FILE WAS GENERATED AUTOMATICALLY BY c2man!
.\" DO NOT EDIT! CHANGES MADE TO THIS FILE WILL BE LOST!
.TH "RuseAs" 3 "13 August 1999" "c2man ruseas.c"
.SH "NAME"
RuseAs \- set the cell representation the application will use
.SH "SYNOPSIS"
.ft B
#include "csf.h"
.br
.sp
int RuseAs
.br
(
.br
	MAP *m,
.br
	CSF_CR useType
.br
);
.ft R
.SH "PARAMETERS"
.TP
.B "MAP *m"
map handle
.TP
.B "CSF_CR useType"
CR_UINT1,CR_INT4, CR_REAL4, CR_REAL8, VS_BOOLEAN or VS_LDD
.sp
Possible values for a \fBCSF_CR\fR are as follows:
.IP
* preferred version 2 cell representations
.RS 0.75in
.PD 0
.ft B
.nr TL \w'CR_UNDEFINED'u+0.2i
.ft R
.TP \n(TLu
\fBCR_UINT1\fR
boolean, ldd and small nominal and small ordinal
.TP \n(TLu
\fBCR_INT4\fR
large nominal and large ordinal
.TP \n(TLu
\fBCR_REAL4\fR
single scalar and single directional
.RE
.PD
.IP
* other version 2 cell representations
.RS 0.75in
.PD 0
.ft B
.nr TL \w'CR_UNDEFINED'u+0.2i
.ft R
.TP \n(TLu
\fBCR_REAL8\fR
double scalar or directional, and also the only type that
can hold all
cell representation without loss of precision
.RE
.PD
.IP
* version 1 cell representations
these can be returned by BUT NOT passed to a csf2 function
.RS 0.75in
.PD 0
.ft B
.nr TL \w'CR_UNDEFINED'u+0.2i
.ft R
.TP \n(TLu
\fBCR_INT1\fR
.
.TP \n(TLu
\fBCR_INT2\fR
.
.TP \n(TLu
\fBCR_UINT2\fR
.
.TP \n(TLu
\fBCR_UINT4\fR
.
.RE
.PD
.IP
* this one CANNOT be returned by NOR passed to a csf2 function 
.RS 0.75in
.PD 0
.ft B
.nr TL \w'CR_UNDEFINED'u+0.2i
.ft R
.TP \n(TLu
\fBCR_UNDEFINED\fR
just some value different from the rest
.RE
.PD
.SH "DESCRIPTION"
RuseAs enables an application to use cell values
in a different format than they are stored in the map.
Cell values are converted when getting (Rget*-functions) and
putting (Rput*-functions) cells if necessary.
Thus no conversions are applied if cell representation and/or
value scale already match.
Any conversions between the version 2 cell representations,
(CR_UINT1, CR_INT4, CR_REAL4 and CR_REAL8) is possible.
Conversion from a non version 2 cell representation to a version
2 cell representation is only possible when you don't
have write access to the cells.
Conversion rules are exactly as described in K&R 2nd edition section A6.

Two special conversions are possible if you don't
have write access to the cells or if the in-file cell representation is
UINT1:
(1) VS_BOOLEAN: successive calls to the Rget*-functions returns the result of
value != 0
, that is 0 or 1 in UINT1 format. The in-file cell representation can be
anything, except if the value scale is VS_DIRECTION or VS_LDD.
(2) VS_LDD: successive calls to the Rget*-functions returns the result of
value % 10
, that is 1 to 9 in UINT1 format (0's are set to MV_UINT1).
The in-file cell representation must be CR_UINT1 or CR_INT2 and
the value scale must be VS_LDD, VS_CLASSIFIED or VS_NOTDETERMINED.
.SH "RETURNS"
0 if conversion obeys rules given here. 1 if not (conversions
will not take place).
.SH "NOTE"
that you must use Rmalloc() to get enough space for both the in-file and
app cell representation.
.SH "MERRNO"
CANT_USE_AS_BOOLEAN CANT_USE_WRITE_BOOLEAN
CANT_USE_WRITE_LDD
CANT_USE_AS_LDD
CANT_USE_WRITE_OLDCR
ILLEGAL_USE_TYPE
.SH "EXAMPLE"
.lf 1 examples/maskdump.tr
.DS
 #include <stdio.h>
 #include <stdlib.h>
 #include "csf.h"
 
 /* a simple csf-mask to stdout
  * program, with minimal checking
  * a mask is boolean-map, here we
  * use version 1 compatibility. 
  */
 
 void main(int argc, char *argv[] )
 {
 
   UINT1 *row;
   MAP *map;                      
   size_t r,c;
 
   if (argc != 2)
   {
    fprintf(stderr,"%s: no file specified\\n",argv[0]);
    exit(1);
   }
 
   map = Mopen(argv[1], M_READ);
 
   if (! RvalueScaleIs(map, VS_BOOLEAN))
   {    /* it's not VS_BOOLEAN, VS_CLASSIFIED 
         * or VS_NOTDETERMINED 
         */
        fprintf(stderr,"'%s' is not a boolean map\\n",argv[1]);
        exit(1);
   }
 
   RuseAs(map, VS_BOOLEAN); 
   row = (UINT1 *)Rmalloc(map, RgetNrCols(map));
 
   for(r=0; r < RgetNrRows(map); r++)
   {
    RgetRow(map, r, row);
    for(c=0; c < RgetNrCols(map); c++)
      printf("%d ",(int)row[c]);
    printf("\\n");
   }
 
   free(row);
   Mclose(map);
 
   exit(0);
 }
 
.DE
.lf 136 RuseAs.3


src/CHANGES
/*
CHANGES log (summary)
============================
Changes written after a version number apply to the next the version.
date: version (program name)

dec 92, spil 2.06:
	added method ossfim, dump_cvv(); dynamic data maximum.
jan 93, spil 2.07:
	dynamic allocation data, unbounded maximum.
mar 93, spil 2.08:
	corrected block discretizing
apr 93, spil 2.09:
	added lex variogram expression parsing (rd_vario.[lc]) ## later removed
apr 93, spil 3.0:
	added full command file parser with lexical analyser
	added sample variogram calculation (sem)

jun 93, gstat 0.99
	spil 3.0 + sem -> gstat 0.99; "gstat [-]sem" ## later: -e semivariance

jun 93, gstat 0.99c:
	removed bug in local/ordinary kriging
jun 93, gstat 0.99d:
	coupled gstat with meschach library; removed numerical recip matrix code.
jun 93, gstat 0.99e:
	solved bug in semi().
aug 93, gstat 0.99f:
	added MapSign().
aug 93, gstat 0.99g:
	added time coordinate to data, distance(x,t) not solved.
	corrected distance bug when d->force in select.c
	added variogram->max_[range/sill] 
aug 93, gstat 0.99h
	added max_ldl structure, and correction with -L option
sep 93, gstat 0.99i
	dynamic allocation of nested variogram models (unbounded maximum)
sep 93, gstat 0.99j
	added universal kriging
oct 93, gstat 1.0  
	release without univ.kr., added variography check (ic,lmc).
nov 93, gstat 1.1  
	renewed error handling, corrected variography check
	corrected bug in memory allocation in ReadLegend();
	released gstat 1.1 on dec 1, 1993

dec 93, gstat 1.2 
	changed the VARIOGRAM structure, using V_model and V_model_type;
	[ check UK, consider REML? ]
	added 4x4 Gauss quadrature block descretizing (Carr & Palmer)
	removed the attr2 field from POINT, changed semi() accordingly
	removed bug in positioning results from ID, if n_vars > 1
	changed UK, added LM. def: with intercept, other variables with X
	revised sem.c , added sem_main.c, made CheckX();
apr 94, gstat 1.2a 
	removed bug in sem_main.c (log, and I for data2);
	removed bug in do_mask.c, warning for coordinates when PT_XY
	removed bug in direct.c (117), added ER_DOMAIN
apr 94, gstat 1.2b
	added gram_sch.c: Gram_Schmidt checking of matrices: option -g
	added # comment-style and flex -l compatability (GNU flex 2.4.6)
	removed bug in header cross semivariogram (pseudo <--> symmetric)
	changed Quant() and Quantile, so they equal Splus' quantile()
june 94, gstat 1.2c
	removed bug when using both Valid and XValid (reversed order)
	removed method tin, lss, ... added method xy
june 94, gstat 1.2d
	removed bug in Skrige() (Oops, I forgot to subtract sk_mean)
june 94, gstat 1.2e
	added reading `points(xxx):"map.csf"'; points from csf map.
    released 1.2f
june 94, gstat 1.2g 
	corrected bug in lm.c: pred. variance, depending on block size
	placed '#' on start of the 8 header lines of semivariograms, 
	to make life easier for gnuplot support
aug 94, gstat 1.2h
	added 'par:fit=1,gnuplot="file.gp"; functionality: see sem.c, 
	sem_fit.c this is for wls fitting of sill-nugget, and variogram 
	plotting using gnuplot. (added # in variogram header!)
aug 94, gstat 1.3 
	Added mivque.[ch], for mi[nv]que univariate covariance fitting
aug 94, gstat 1.3a
	added global kriging in [s]krige.c. Up to 80 times faster. Wow.
aug 94, gstat 1.3b
	added `area: "discr.dat",x=1,y=2', user defined area discretising
sep 94, gstat 1.3c
	fixed bug in CovSem_AB(): remember both within-block cov. and sem.
oct 94, gstat 1.3d
	added menu; corrected CheckX() for points();, but see 1.4b...
	added noaverage/average instead of "average=1"
	added "square" to use square neigborhood selections (no circle)
	changed variogram structure, now functions as part of structure.
	removed bug in getest, functions for MVS/MED, added par:quantile=..
dec 94, gstat 1.3e 
	menu evolved... added '#','?','-','&' options (try ?);
	added recalculation only at options Data, Calculate, cUtoff, &

----------gstat, 1995:
jan 94, gstat 1.4  
	menu.c elm-like, major rewrite. No hidden options, help added.  
jan 94, gstat 1.4a 
	corrected compile with STRICT_KR2 (no HAVE_CURSES)
	menu faster (incomplete screen rewrite after each key pressed)
	env. variable GNUPROLOG may define a prolog for gnuplot
jan 94, gstat 1.4b 
	corrected bug in select.c (cs with no range defined)
	env. variable GNUPLOT points gnuplot out, GSTATFMT "%g" variogram
	menu enhancements (echo Yes/No, Show)
	Yet another nasty bug in CheckX()!
feb 94, gstat  1.4c 
	header of variogram cloud corrected. Found bug in mopen.c (csf-lib)
	IsEas(): now use rewind() instead of fopen()+fclose().
	partial fitting: 0.2 @Nug() + 1 Sph(30) will now only fit Sph part.
	(removed obsolete Wls_fit_Sill())
	-L : allow local residuals.
	added dX to DATA: max X-distance allowed for semivar. calculation
	(removed obsolete `factor' from DATA. This can now be done
	in a more general fashion with X and dX).
	menu options: 'J': jgraph file; `&' jgraph-eps; '!': shell command.
Mar 94, gstat 1.4d 
	cleaned up random.c, removed rn.[ch] and Ran[0-3]()
	corrected estimation of sample variance (lm->s2) for weighted lm.
 	EnterData() (menu.c) now does not return without valid data read.
	temporary file names now come from tmpnam() or tempnam() [on UNIX];
	added MODE mode: more flexibility now (stratify whenever suitable)
	Fixed bug in Xvalid(): where.X, NA's on uk/lm for secondary vars
	Fixed bug (weights) in CalcLM(); added 's', and '+/-' in menu.
	Added DATE in date.h, changed program heading: -V now includes date.
	Now mean variances (V=..) are subtracted in variogram calculation.
Apr 94, gstat 1.4.e
	Made lm->SSR the corrected SS for mean, with p-1 df (Christensen p. 89)
	Added more.c (callable from menu). (removed later on...)
	syntax change: `pars:x=value;' to `set x=value;' (like gnuplot)
	Made `width' and `cutoff' global variables (`set cutoff=15,width=1').
	fit.c: now use lm.c routines for the fitting (estimation),
	Changed convergence criterion in WlsFit(), Mivque():

libcsf/src/ruseas.c
#ifndef lint
 static const char *rcs_id = 
 "$Header: /home/pcrtree/SRC.RCS/libs/csf/RCS/ruseas.c,v 2.1 1996/12/29 19:35:21 cees Exp $";
#endif

#include "csf.h"
#include "csfimpl.h"

static void UINT1tLdd(size_t nrCells, void *Buf)
{
	size_t i;
	UINT1 *buf = (UINT1 *)Buf;

	for(i=0; i < (size_t)nrCells; i++)
		if (buf[i] != MV_UINT1)
		{
			buf[i] %= (UINT1)10;
			if (buf[i] == 0)
				buf[i] = MV_UINT1;
		}
}

static void INT2tLdd(size_t nrCells, void *Buf)
{
	size_t i;
	INT2  *inBuf  = (INT2  *)Buf;
	UINT1 *outBuf = (UINT1 *)Buf;
	for(i=0; i < (size_t)nrCells; i++)
		if (inBuf[i] != MV_INT2)
		{
			outBuf[i] = (UINT1)(ABS(inBuf[i]) % 10);
			if (outBuf[i] == 0)
				outBuf[i] = MV_UINT1;
		}
		else
			outBuf[i] = MV_UINT1;
}

#define TOBOOL(nr, buf, srcType)\
{\
	size_t i;\
/* loop must be upward to prevent overwriting of values \
 * not yet converted \
 */ \
 	PRECOND(sizeof(srcType) >= sizeof(UINT1));\
	for(i=0; i < (size_t)nr; i++)\
		if (IS_MV_##srcType( ((srcType *)buf)+i) )\
			((UINT1 *)buf)[i] = MV_UINT1;\
		else\
			((UINT1 *)buf)[i] = ((srcType *)buf)[i] != ZERO_##srcType;\
}

static void INT1tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, INT1); }

static void INT2tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, INT2); }

static void INT4tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, INT4); }

static void UINT1tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, UINT1); }

static void UINT2tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, UINT2); }

static void UINT4tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, UINT4); }

static void REAL4tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, REAL4); }

static void REAL8tBoolean(size_t nrCells, void *buf)
{ TOBOOL(nrCells, buf, REAL8); }

#define CONV_BIG_TO_SMALL(nr, buf, destType, srcType)\
{\
	size_t i;\
/* loop must be upward to prevent overwriting of values \
 * not yet converted: \
 */ \
 	PRECOND(sizeof(srcType) >= sizeof(destType)); /* upward loop OK */ \
	for(i=0; i < (size_t)nr; i++)\
		if (IS_MV_##srcType( ((srcType *)buf)+i) )\
		    SET_MV_##destType( ((destType *)buf)+i);\
		else\
			((destType *)buf)[i] = (destType)(((srcType *)buf)[i]);\
}

#define CONV_SMALL_TO_BIG(nr, buf, destType, srcType)\
{\
	size_t i = (size_t)nr;\
/* loop must be downward to prevent overwriting of values \
 * not yet converted: \
 */ \
 	PRECOND(sizeof(srcType) <= sizeof(destType)); /* downward loop OK */ \
	do { i--;\
		if (IS_MV_##srcType( ((srcType *)buf)+i) )\
		    SET_MV_##destType( ((destType *)buf)+i);\
		else\
			((destType *)buf)[i] = (destType)(((srcType *)buf)[i]);\
	}while ( i != 0);\
}

static void UINT1tINT4(size_t nrCells, void *buf)
{ CONV_SMALL_TO_BIG(nrCells, buf, INT4, UINT1);}

static void UINT1tREAL4(size_t nrCells, void *buf)
{ CONV_SMALL_TO_BIG(nrCells, buf, REAL4, UINT1);}

static void UINT1tREAL8(size_t nrCells, void *buf)
{ CONV_SMALL_TO_BIG(nrCells, buf, REAL8, UINT1);}

static void INT4tUINT1(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, UINT1, INT4);}

static void INT2tUINT1(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, UINT1, INT2);}

static void UINT2tUINT1(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, UINT1, UINT2);}

static void INT4tREAL4(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, REAL4, INT4);}

static void INT4tREAL8(size_t nrCells, void *buf)
{ CONV_SMALL_TO_BIG(nrCells, buf, REAL8, INT4);}

static void REAL4tUINT1(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, UINT1, REAL4);}

static void REAL4tINT4(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, INT4, REAL4);}

static void REAL4tREAL8(size_t nrCells, void *buf)
{ CONV_SMALL_TO_BIG(nrCells, buf, REAL8, REAL4);}

static void REAL8tUINT1(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, UINT1, REAL8);}

static void REAL8tINT4(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, INT4, REAL8);} 

static void REAL8tREAL4(size_t nrCells, void *buf)
{ CONV_BIG_TO_SMALL(nrCells, buf, REAL4, REAL8);}

static void Transform2( size_t nrCells, void *buf, CSF_CR destCellRepr, CSF_CR currCellRepr);

static void INT1tINT4(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_INT4, CR_INT1);}

static void INT1tREAL4(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL4, CR_INT1);}

static void INT1tREAL8(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL8, CR_INT1);}

static void INT2tINT4(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_INT4, CR_INT2);}

static void INT2tREAL4(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL4, CR_INT2);}

static void INT2tREAL8(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL8, CR_INT2);}

static void UINT2tINT4(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_INT4, CR_UINT2);}

static void UINT2tREAL4(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL4, CR_UINT2);}

static void UINT2tREAL8(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL8, CR_UINT2);}

static void UINT4tREAL4(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL4, CR_UINT4);}

static void UINT4tREAL8(size_t nrCells, void *buf)
{ Transform2(nrCells, buf, CR_REAL8, CR_UINT4);}


static void ConvertToINT2( size_t nrCells, void *buf, CSF_CR src)
{
	if (IS_SIGNED(src))
	{
		POSTCOND(src == CR_INT1);
		CONV_SMALL_TO_BIG(nrCells,buf, INT2, INT1);
	}
	else  
	{
		POSTCOND(src == CR_UINT1);
		CONV_SMALL_TO_BIG(nrCells,buf, INT2, UINT1);
	}
}

static void ConvertToINT4( size_t nrCells, void *buf, CSF_CR src)
{
	if (IS_SIGNED(src))
	{
		POSTCOND(src == CR_INT2);
		CONV_SMALL_TO_BIG(nrCells,buf, INT4, INT2);
	}
	else  
	{
		POSTCOND(src == CR_UINT2);
		CONV_SMALL_TO_BIG(nrCells,buf, INT4, UINT2);
	}
}


static void UINT1tUINT2(
size_t nrCells,
void *buf)
{
	CONV_SMALL_TO_BIG(nrCells, buf, UINT2, UINT1);
}

static void UINT2tUINT4(
size_t nrCells,
void *buf)     
{
	CONV_SMALL_TO_BIG(nrCells, buf, UINT4, UINT2);
}

static void ConvertToREAL4( size_t nrCells, void *buf, CSF_CR src)
{
	size_t i;

	i = (size_t)nrCells;

	if (IS_SIGNED(src))
	{
		POSTCOND(src == CR_INT4);
		INT4tREAL4(nrCells, buf);
	}
	else
	{
		POSTCOND(src == CR_UINT4);
		{
			do {
				i--;
				if ( ((UINT4 *)buf)[i] == MV_UINT4 )
					((UINT4 *)buf)[i] = MV_UINT4;
				else
					((REAL4 *)buf)[i] = ((UINT4 *)buf)[i];
			    }
			while(i != 0);
		}
	}
}

static void Transform2(
	size_t nrCells,
	void  *buf,
	CSF_CR destCellRepr, /* the output representation */
	CSF_CR currCellRepr)  /* at start of while this is the representation
				read in the MAP-file */
{
  /* subsequent looping changes the to the new
   * converted type
   */
	while(currCellRepr != destCellRepr)
	{
		switch(currCellRepr)
		{
			case CR_INT1:    ConvertToINT2(nrCells, buf,
						currCellRepr);
					currCellRepr = CR_INT2;
					break;
			case CR_INT2:    ConvertToINT4(nrCells, buf, 
						currCellRepr);
					currCellRepr = CR_INT4;
					break;
			case CR_INT4:    ConvertToREAL4(nrCells, buf,
						currCellRepr);
					currCellRepr = CR_REAL4;
					break;
			case CR_UINT1: 	if (IS_SIGNED(destCellRepr))
					{
						ConvertToINT2(nrCells, buf,
							currCellRepr);
						currCellRepr = CR_INT2;
					}
					else
					{
						UINT1tUINT2(nrCells, buf);
						currCellRepr = CR_UINT2;
					}
					break;
			case CR_UINT2:   if (destCellRepr == CR_INT4)
					{
						ConvertToINT4(nrCells, buf,
							currCellRepr);
					 	currCellRepr = CR_INT4;
					 }
					 else
					 {
						UINT2tUINT4(nrCells, buf);
						currCellRepr = CR_UINT4;
					 }
					 break;
			case CR_UINT4:   ConvertToREAL4(nrCells, buf,
						currCellRepr);
					currCellRepr = CR_REAL4;
					break;
			default       :	POSTCOND(currCellRepr == CR_REAL4);
					REAL4tREAL8(nrCells, buf);
					currCellRepr = CR_REAL8;
		}
	}
}

/* OLD STUFF 
void TransForm(
	const MAP *map,
	size_t nrCells,
	void *buf)
{
	Transform2(nrCells, buf, map->types[READ_AS], map->types[STORED_AS]);
}
*/

#define illegal NULL
#define same    CsfDummyConversion

static const CSF_CONV_FUNC ConvTable[8][8] = {
/* ConvTable[source][destination] */
/* INT1     , INT2      , INT4      , UINT1     , UINT2     , UINT4     , REAL4     , REAL8        */
/* 0        , 1         , 2         , 3         , 4         , 5         , 6         , 7      ind   */
/* 0x04     , 0x05      , 0x06      , 0x00      , 0x01      , 0x02      , 0x0A      , 0x0B low-nib */
{  same     , illegal   , INT1tINT4 , illegal   , illegal   , illegal   , INT1tREAL4, INT1tREAL8}, /* INT1 */
{ illegal   ,  same     , INT2tINT4 ,INT2tUINT1 , illegal   , illegal   , INT2tREAL4, INT2tREAL8}, /* INT2 */
{ illegal   , illegal   ,  same     ,INT4tUINT1 , illegal   , illegal   , INT4tREAL4, INT4tREAL8}, /* INT4 */
{ illegal   , illegal   ,UINT1tINT4 ,  same     ,UINT1tUINT2, illegal   ,UINT1tREAL4,UINT1tREAL8}, /* UINT1 */
{ illegal   , illegal   ,UINT2tINT4 , UINT2tUINT1 ,  same     ,UINT2tUINT4,UINT2tREAL4,UINT2tREAL8}, /* UINT2 */
{ illegal   , illegal   , illegal   , illegal   , illegal   ,  same     ,UINT4tREAL4,UINT4tREAL8}, /* UINT4 */
{ illegal   , illegal   ,REAL4tINT4 ,REAL4tUINT1, illegal   , illegal   ,  same     ,REAL4tREAL8}, /* REAL4 */
{ illegal   , illegal   ,REAL8tINT4 ,REAL8tUINT1, illegal   , illegal   ,REAL8tREAL4,  same     }  /* REAL8 */
};

static const CSF_CONV_FUNC boolConvTable[8] = {
INT1tBoolean, INT2tBoolean, INT4tBoolean,
UINT1tBoolean, UINT2tBoolean, UINT4tBoolean,
REAL4tBoolean, REAL8tBoolean };


static const char  convTableIndex[12] = {
	   3 /* UINT1 */,  4 /* UINT2 */,  5 /* UINT4 */, -1 /* 0x03  */,
	   0 /*  INT1 */,  1 /*  INT2 */,  2 /*  INT4 */, -1 /* 0x07  */,
	  -1 /*  0x08 */, -1 /*  0x09 */,  6 /* REAL4 */,  7 /* REAL8 */
};

static CSF_CONV_FUNC ConvFuncBool(CSF_CR cr)
{
	PRECOND(CSF_UNIQ_CR_MASK(cr) < 12);
	PRECOND(convTableIndex[CSF_UNIQ_CR_MASK(cr)] != -1);
	
	return boolConvTable[(int)convTableIndex[CSF_UNIQ_CR_MASK(cr)]];
}

static CSF_CONV_FUNC ConvFunc(CSF_CR destType, CSF_CR srcType)
{

	PRECOND(CSF_UNIQ_CR_MASK(destType) < 12);
	PRECOND(CSF_UNIQ_CR_MASK(srcType) < 12);
	PRECOND(convTableIndex[CSF_UNIQ_CR_MASK(srcType)] != -1);
	PRECOND(convTableIndex[CSF_UNIQ_CR_MASK(destType)] != -1);
	/* don't complain on illegal, it can be attached
	 * to a app2file while there's no WRITE_MODE
	 * if it's an error then it's catched in RputSomeCells
	 */
	return 
         ConvTable[(int)convTableIndex[CSF_UNIQ_CR_MASK(srcType)]]
 	          [(int)convTableIndex[CSF_UNIQ_CR_MASK(destType)]];
}

static int HasInFileCellReprType2(CSF_CR cr)
{
	char  type2[12] = {
	   1 /* UINT1 */,  0 /* UINT2 */,  0 /* UINT4 */,  0 /* 0x03  */,
	   0 /*  INT1 */,  0 /*  INT2 */,  1 /*  INT4 */,  0 /* 0x07  */,
	   0 /*  0x08 */,  0 /*  0x09 */,  1 /* REAL4 */,  1 /* REAL8 */};

	PRECOND(CSF_UNIQ_CR_MASK(cr) < 12);
	
	return (int)type2[CSF_UNIQ_CR_MASK(cr)];
}

/* set the cell representation the application will use
 * RuseAs enables an application to use cell values
 * in a different format than they are stored in the map.
 * Cell values are converted when getting (Rget*-functions) and
 * putting (Rput*-functions) cells if necessary.
 * Thus no conversions are applied if cell representation and/or 
 * value scale already match.
 * Any conversions between the version 2 cell representations, 
 * (CR_UINT1, CR_INT4, CR_REAL4 and CR_REAL8) is possible. 
 * Conversion from a non version 2 cell representation to a version
 * 2 cell representation is only possible when you don't
 * have write access to the cells.
 * Conversion rules are exactly as described in K&R 2nd edition section A6.
 * 
 * Two special conversions are possible if you don't
 * have write access to the cells or if the in-file cell representation is
 * UINT1:
 * (1) VS_BOOLEAN: successive calls to the Rget*-functions returns the result of
 * value != 0 
 * , that is 0 or 1 in UINT1 format. The in-file cell representation can be
 * anything, except if the value scale is VS_DIRECTION or VS_LDD.
 * (2) VS_LDD: successive calls to the Rget*-functions returns the result of
 * value % 10 
 * , that is 1 to 9 in UINT1 format (0's are set to MV_UINT1). 
 * The in-file cell representation must be CR_UINT1 or CR_INT2 and 
 * the value scale must be VS_LDD, VS_CLASSIFIED or VS_NOTDETERMINED.
 *
 * NOTE
 * that you must use Rmalloc() to get enough space for both the in-file and
 * app cell representation. 
 *
 * returns 
 * 0 if conversion obeys rules given here. 1 if not (conversions 
 * will not take place).
 *
 * Merrno
 * CANT_USE_AS_BOOLEAN CANT_USE_WRITE_BOOLEAN
 * CANT_USE_WRITE_LDD
 * CANT_USE_AS_LDD
 * CANT_USE_WRITE_OLDCR
 * ILLEGAL_USE_TYPE
 *
 * EXAMPLE
 * .so examples/maskdump.tr
 */

int RuseAs(
	MAP *m,          /* map handle */
	CSF_CR useType)   /* CR_UINT1,CR_INT4, CR_REAL4, CR_REAL8, VS_BOOLEAN or VS_LDD */
{ 

  CSF_CR inFileCR = RgetCellRepr(m);
  CSF_VS inFileVS = RgetValueScale(m);
  int hasInFileCellReprType2 =  HasInFileCellReprType2(inFileCR);

  switch(useType)
  {
    case VS_BOOLEAN:
  	switch(inFileVS) {
	   case VS_LDD: case VS_DIRECTION:
   			M_ERROR(CANT_USE_AS_BOOLEAN);
   			return 1;
   	   case VS_BOOLEAN:
   	   		POSTCOND(inFileCR == CR_UINT1);
   	   		m->appCR = CR_UINT1;
   	   		m->file2app = same;
   	   		m->app2file = same;
   	   		return 0;
   	   default:
			if ( (! hasInFileCellReprType2) && WRITE_ENABLE(m) )
			{ /* cellrepr is old one, we can't write that */
					M_ERROR(CANT_USE_WRITE_BOOLEAN);
					return 1;
			}
   			m->appCR = CR_UINT1;
   			m->file2app  = ConvFuncBool(inFileCR);
   			m->app2file = ConvFunc(inFileCR, CR_UINT1);
   	   		return 0;
   	} /* case useType == VS_BOOLEAN */
   	break;

   case VS_LDD:
        switch(inFileVS) {
         case VS_LDD:
   	   		POSTCOND(inFileCR == CR_UINT1);
   	   		m->appCR = CR_UINT1;
   	   		m->file2app = same;
   	   		m->app2file = same;
   	   		return 0;
        case VS_CLASSIFIED: 
        case VS_NOTDETERMINED: 
        	switch(inFileCR) {
        	 case CR_UINT1:
        	 	m->appCR = CR_UINT1;
			m->file2app  = UINT1tLdd;
			m->app2file = same;
			return 0;
		 case CR_INT2:
		        if (WRITE_ENABLE(m))
		        {  M_ERROR(CANT_USE_WRITE_LDD);
			   return 1;
		        }
        	 	m->appCR = CR_UINT1;
			m->file2app  = INT2tLdd;
			m->app2file = illegal;
			return 0;
		}
	default: M_ERROR(CANT_USE_AS_LDD);
		 return 1;
      }
      /* case useType == VS_LDD */
      break;
     case CR_UINT1:
     case CR_INT4 :
     case CR_REAL4:
     case CR_REAL8:
		if ( (! hasInFileCellReprType2) && WRITE_ENABLE(m) )
		{ /* cellrepr is old one, we can't write that */
				M_ERROR(CANT_USE_WRITE_OLDCR);
				return 1;
		}
		m->appCR = useType;
		m->file2app  = ConvFunc(useType, inFileCR);
		m->app2file = ConvFunc(inFileCR, useType);
		POSTCOND(m->file2app != NULL);
		return 0;
    default:
    		M_ERROR(ILLEGAL_USE_TYPE);
		return 1;
  }
  /* NOTREACHED */
}
